Automattic/woocommerce-subscriptions-core
# Fix duplicate related order caches after HPOS post backfill

## Description

Resolves issues raised in https://github.com/Automattic/woocommerce-subscriptions-core/pull/538#issuecomment-1801122335

WC core includes a bug (https://github.com/woocommerce/woocommerce/pull/41281) that has two affects when attempting to backfill subscription related order meta. If the cache is empty it wouldn't copy the meta at all. If the cache had values, it would duplicate. The missing meta was resolved by #538 however that meant it was now always duplicating the meta on the latest released version of WC. 

I've submitted a fix to WC core (https://github.com/woocommerce/woocommerce/pull/41281), however while we wait for that to be released we want to prevent those duplicate cache meta values from appearing after a backfill. 

This PR fixes that by making sure we delete all instances of a subscriptions related order cache after the WC part of the backfill is completed before we resave a final-single copy of the cache.

The flow is as follows:

1. WC duplicates the meta in `update_order_meta_from_object()` called [here](https://github.com/woocommerce/woocommerce/blob/797653c334c4d123b582fe4ab3a5a81a70de74b9/plugins/woocommerce/includes/data-stores/abstract-wc-order-data-store-cpt.php#L663)
    - see the PR to WC core for additional details - https://github.com/woocommerce/woocommerce/pull/41281.
3. WC core [calls the datastore getters and setters](https://github.com/woocommerce/woocommerce/blob/d83c7f30f7a60f4b8955daae169e48e3301e696c/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php#L594-L610).
4. That calls the functions I've changed in this PR. 
5. We delete all duplicates previously added by core and then save a single copy of the cache. 







## How to test this PR

1. Make sure you're using the latest version of WC - 8.2.1
1. Go **WC > Settings > Advanced > Features** and enable HPOS without compatibility mode
2. Purchase 2 different subscription products (we want 2 subscriptions). 
3. Populate the renewal order IDs cache meta by processing a renewal order for 1 of the  subscriptions
4. Open the HPOS settings again and enable compatibility mode and save the settings
5. If the sync process is slow to kick-off here are some steps to manually run it:
   1. Run the `wc_schedule_pending_batch_processes` pending scheduled action
   2. Then run the new `wc_run_batch_process` scheduled action
6. Open your DB and search `wp_postmeta` for the subscription ID
   - On `fix/hpos-dates-cache-not-backfilled` you should see duplicate related order caches. 
   - This branch there should be no duplicates. 

| `fix/hpos-dates-cache-not-backfilled` | This Branch |
|:--------:|:--------:|
| <img width="590" alt="Screenshot 2023-11-08 at 4 54 35 pm" src="https://github.com/Automattic/woocommerce-subscriptions-core/assets/8490476/806e20ed-491e-4fc8-93a0-756550b42e28"><br><sup>Duplicates</sup> | <img width="617" alt="Screenshot 2023-11-08 at 4 51 52 pm" src="https://github.com/Automattic/woocommerce-subscriptions-core/assets/8490476/f1dc1443-a05e-457c-b7e8-94aa9baac84f"><br><sup>No duplicates</sup>|





## Product impact


   - N/A
